ecommerce_integrations/shopify/product.py

```diff
@@ -342,51 +342,125 @@ def upload_erpnext_item(doc, method=None):
 	if frappe.flags.in_import:
 		return
 
-	if doc.has_variants or doc.variant_of:
-		msgprint(_("Item with variants or template items can not be uploaded to Shopify."))
+	if item.has_variants:
+		# msgprint(_("Template items can not be uploaded to Shopify."))
 		return
 
+	if doc.variant_of and not setting.upload_variants_as_items:
+		msgprint(_("Enable variant sync in setting to upload item to Shopify."))
+		return
+
+	if item.variant_of:
+		template_item = frappe.get_doc("Item", item.variant_of)
+
 	product_id = frappe.db.get_value(
 		"Ecommerce Item",
-		{"erpnext_item_code": item.name, "integration": MODULE_NAME},
+		{"erpnext_item_code": template_item.name, "integration": MODULE_NAME},
 		"integration_item_code",
 	)
 	is_new_product = not bool(product_id)
 
 	if is_new_product:
-
 		product = Product()
 		product.published = False
-		product.status = "draft"
+		product.status = "draft" if setting.sync_new_item_as_draft else "active"
 
-		map_erpnext_item_to_shopify(shopify_product=product, erpnext_item=item)
+		map_erpnext_item_to_shopify(shopify_product=product, erpnext_item=template_item)
 		is_successful = product.save()
 
 		if is_successful:
 			update_default_variant_properties(
-				product, sku=item.item_code, price=item.standard_rate, is_stock_item=item.is_stock_item,
+				product,
+				sku=template_item.item_code,
+				price=template_item.standard_rate,
+				is_stock_item=template_item.is_stock_item,
 			)
+			if item.variant_of:
+				product.options = []
+				product.variants = []
+				variant_attributes = {"title": template_item.item_name}
+				max_index_range = min(3, len(template_item.attributes))
+				for i in range(0, max_index_range):
+					attr = template_item.attributes[i]
+					product.options.append(
+						{
+							"name": attr.attribute,
+							"values": frappe.db.get_all(
+								"Item Attribute Value", {"parent": attr.attribute}, pluck="attribute_value"
+							),
+						}
+					)
+					try:
+						variant_attributes[f"option{i+1}"] = item.attributes[i].attribute_value
+					except IndexError:
+						frappe.throw(_("Shopify Error: Missing value for attribute {}").format(attr.attribute))
+				product.variants.append(Variant(variant_attributes))
+
 			product.save()  # push variant
 
-			ecom_item = frappe.get_doc(
-				{
-					"doctype": "Ecommerce Item",
-					"erpnext_item_code": item.name,
-					"integration": MODULE_NAME,
-					"integration_item_code": str(product.id),
-					"variant_id": str(product.variants[0].id),
-					"sku": str(product.variants[0].sku),
-				}
-			)
-			ecom_item.insert()
+			ecom_items = list(set([item, template_item]))
+			for d in ecom_items:
+				ecom_item = frappe.get_doc(
+					{
+						"doctype": "Ecommerce Item",
+						"erpnext_item_code": d.name,
+						"integration": MODULE_NAME,
+						"integration_item_code": str(product.id),
+						"variant_id": "" if d.has_variants else str(product.variants[0].id),
```

Comment:
They ALL have same variant ID that's why inventory sync behaviour is buggy

![image](https://user-images.githubusercontent.com/9079960/199172856-fad5e6d6-4623-47e2-a2c4-f12f931d5741.png)



At least test once!

Reply:
Shopify seem to not have a fixed order for the variants. After each `product.save()`, the order of the list seems to mutate.
I had limited number of attribute values and unfortunately, new variants came as the last ones when I was writing the logic and I decided to go with `-1` index.

Now that I check with more data, it breaks. I have written loop to check for matching conditions and then set the variant ID.